velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,530 @@
+cardinality workloads (large number of groups) the size of the hash table
+exceeds the query’s memory limit.
+
+To allow such queries to success, spilling kicks in and writes part of the
+operator's state to disk. After operator received all input, it reads back the
+spilled state from disk and merges it with un-spilled state in memory to
+produce the result. The spilling process consists of two phases: spilling and
+restoring. The spilling phase is executed while the operator processes input.
+It decides which parts of the operator state to spill and how to store the
+spilled state on disk. The restoring phase is executed after the operator has
+processed all the input. It reads the spilled state from disk and merges it
+with un-spilled state in memory to produce the result. Different operators use
+different spilling algorithms. This document discusses the algorithms used by
+Hash Aggregation, Order By and Hash Join operators.
+
+Spilling Framework
+------------------
+
+Spilling in Velox consists of a spilling framework and a set of spillable
+operators built on top of it. The spilling framework provides the common
+spilling functions which includes spill data collection, partition, sort,
+serialization/deserialization and the storage read and write. Each spillable
+operator uses these functions to implement its own spilling algorithm.
+
+.. image:: images/spill-arch.png
+    :width: 300
+    :align: center
+
+Spill Objects
+-------------
+
+The spilling framework consists of the following major software objects:
+
+`Spiller <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spiller.h#L24>`_
+^^^^^^^
+The Spiller object provides the spilling functions for all the operators
+and helps them manage their spilled state on disk. There is one Spiller object
+created for each operator. The Spiller object takes a row container object on
+construction (except hash probe operator which doesn’t buffer input and
+directly writes spilling rows to disk). The row container is a row-wise
+in-memory data store owned by the operator to store its in-memory state that
+could be spilled to disk when there is a lack of memory. For example, the row
+container of a hash aggregation operator stores the intermediate aggregation
+state with one row per each group (a unique combination of the values of the
+grouping keys). When the operator needs to spill, the Spiller scans all the
+rows in the row container and calculates each row’s partition number, decides
+which partitions need to spill, and then write them out as a list of files. The
+Spiller can sort the data before writing to disk. If sorting is enabled, the
+Spiller creates a separate file for each sorted run. On restore, the Spiller
+reads back the spilled data to restore the in-memory state. If sorting is
+enabled, it creates a merge sort reader to read the sorted data. This
+functionality is used by Hash Aggregation and Order By operators.
```

Comment:
Because we need to merge multiple intermediate states of the same group into the final aggregation state. During the hash aggregation processing, a group state could be spilled multiple times.